### ISSUE TITLE
Adding 'extra' authors to the wp_term_taxononmy

### ISSUE THREAD

**Issue by phylaxis** [Automattic/Co-Authors-Plus#486]
I've been testing this plug-in in a development environment to see if we want to use it and I have a question/concern about something I'm seeing in the database. It seems to be adding entries to the `wp_term_taxononmy` table even if all I am doing is searching for an author in the admin interface. Is that "normal" behavior?

**Comment by phylaxis** [comment]
Hey, just checking in on this it's been about two weeks. Am I the only one noticing this behavior? Could it possibly be a conflict with another add-on or this it just expected behavior go the Co-Authors Plus plugin?

**Comment by philipjohn** [comment]
Hi! CAP uses terms to map "guest authors" to real WP users, so yes, it does add to the terms table.

Additions don't happen when _searching_ though, only when explicitly adding new guest authors.

**Comment by phylaxis** [comment]
But @philipjohn that's exactly the behavior I'm seeing which is why I think there might be a bug or conflict. As I'm searching for a co-authors it is actively adding records to the `wp_term_taxononmy` table. I just tested it in my test environment again and just searching to add a co-author added 24 records to my `wp-term-taxonomy` table

**Comment by phylaxis** [comment]
For what it's worth, I just tested in our environment after disabling ALL plug-ins except CAP and still had records added to the `wp_term_taxonomy` table with just a search alone.

**Comment by phylaxis** [comment]
Hey @philipjohn I'm not sure how to re-open a case, but I can confirm almost definitively that there is a bug somewhere, at least under certain circumstances, where CAP injects records into the `wp_term_taxononmy` table just on a search.

The insert is happening on line 1385 of the co-authors-plus.php file:
```
$new_term = wp_insert_term( $coauthor->user_login, $this->coauthor_taxonomy, $args );
```

I'm seeing that called when the `update_author_term()` function is called inside the loop starting at line 1110 of the `search_authors()` function.

```
foreach ( $found_users as $found_user ) {
   $term = $this->get_author_term( $found_user );
      if ( empty( $term ) || empty( $term->description ) ) {
         $this->update_author_term( $found_user );
      }
  }
```

What I cannot figure out is "why" the if statement at line 1373 is failing and causing the search to trigger the insert at line 1385.

**Comment by rebeccahum** [comment]
Hey @phylaxis, can you please provide the exact steps to reproduce this?  Also, are you able to reproduce this issue on another WP install?

**Comment by phylaxis** [comment]
The issue could be VERY hard to test on another install of Wordpress because I think the bug relates to something with the `wp_users` table and how entries are in there. My suscpicion was it could be related to the same issue that is causing this reported bug, https://github.com/Automattic/Co-Authors-Plus/issues/427, the user_login being different from the user_nicename. We allow our users to register with a username or email address. When the `user_login` is and email address it records the `user_nicename` as the email with the . and @ replaced with dashes. So the `user_login` say `joe@cool.com` would have a `user_nicename` of `joe-cool-com`.

With all this in mind I tried setting up a fresh install of Wordpress 4.9.5 and only Co-Author Plus 3.3.2 installed. I then imported only the `wp_user` table from my database, about 2,400 users. When I did a search to add a co-author after the AJAX ran and without adding the co-author it had inserted 11 records into the `wp_term_taxonomy` table.

Here are the steps to reproduce:

1) Fresh install of Wordpress 
2) `wp_users` table with a significant sample set.
3) Install and activate Co-Authors Plus.
4) Go into the `Hello World` post and edit it.
5) Search for an author in Co-authors plus, but don't select or add them.
6) Look in the `wp_term_taxonomy` table and confirm records have been inserted.

After doing my test I did notice that ALL the user records that had matches for the name I seached for, in this case `dave`, had matching records added to the `wp_term_taxonomy` table. So that might rule out my theory about this being related to an issue where `user_login` and `user_nicename` don't match.

**Comment by phylaxis** [comment]
OK. I did a smaller test and and can confirm that this is probably just a straight-up bug with how the author search is working. With just 10 user records and some with the `user_login` being the same as the `user_nicename` and some not I repeated the steps I outlined. When I did the test all 10 users had matching records inserted into the `wp_term_taxonomy` table.

**Comment by phylaxis** [comment]
Hey @rebeccahum just circling back since I hadn't heard anything back, was this information helpful?

**Comment by TheCrowned** [comment]
Hi @phylaxis,
so I think you are right, and it is a feature indeed. Co-Authors Plus relies on wp_terms to search for co-authors. 

Suppose you installed the plugin on a website that already has some users. When you search for co-authors to be added to a post, the plugin first performs a search among regular WordPress users, looking for your keywords in the following fields: `user display name, user ID, username, user email`. All the users that are found result then in the addition of a new wp_terms entry (unless, of course, that does not already exist), that is used for future queries.

Hope this helps in explaining the behavior just a bit more! :)

Have a nice day,
Stefano

**Comment by phylaxis** [comment]
I'd be with you on this @TheCrowned , but we have a site with thousands of "subscribers" and have no interest in ever making them "co-authors", so this feature is in a way "polluting" the database. Is there a possibility of adding the ability to "disable" this feature? Seems like a simple search should not be inserting records into the DB in ALL cases. Or better yet, filter the inserts to only include users at a specific Wordpress permission level or higher?

**Comment by TheCrowned** [comment]
As you have seen, this is a problem that only shows up when you install the plugin on a website that already has some users. The plugin needs to create wp_terms for all users that may ever be added as coauthors for quicker search, but it is true that it should not happen on search, and that subscribers should be excluded: I agree on this. 

So, the plan we would like to put in action is to **have a cron job** created on plugin activation. This will run in background in wp-admin, slowly creating apt records for all needed users. We would still want to exclude subscribers and any user who could not be added as a coauthor, so we will develop with this in mind. The search function should also get quicker.

**Comment by phylaxis** [comment]
Hey @TheCrowned thanks for being open to my feedback. If you need any help testing this as you develop the feature let me know.

**Comment by TheCrowned** [comment]
Hi @phylaxis! I should have implemented everything that we have talked about :)

I have tested it and it all seems to be working fine for me. I have also added wp_terms deletion for users that should not have a wp_term set up, but that ended up having it with the old code. I will write a detailed PR with all the changes also for you to have a look to, but if you want to test this out in the meantime, checkout this branch https://github.com/TheCrowned/Co-Authors-Plus/tree/feature/486-cron-create-terms :)
